app/src/schemas/toolCallSchemas.ts

```diff
@@ -0,0 +1,189 @@
+ *   }
+ * }
+ * ```
+ */
+export type OpenAIToolCall = z.infer<typeof openAIToolCallSchema>;
+
+/**
+ * The zod schema for multiple OpenAI Tool Calls
+ */
+export const openAIToolCallsSchema = z.array(openAIToolCallSchema);
+
+/**
+ * The JSON schema for multiple OpenAI tool calls
+ */
+export const openAIToolCallsJSONSchema = zodToJsonSchema(
+  openAIToolCallsSchema,
+  {
+    removeAdditionalStrategy: "passthrough",
+  }
+);
+
+/**
+ * The schema for an Anthropic tool call, this is what a message that calls a tool looks like
+ */
+export const anthropicToolCallSchema = z
+  .object({
+    id: z.string().describe("The ID of the tool call"),
+    type: z.literal("tool_use"),
+    name: z.string().describe("The name of the tool"),
+    input: z.record(z.unknown()).describe("The input for the tool"),
+  })
+  .passthrough();
+
+/**
+ * The type of an Anthropic tool call
+ */
+export type AnthropicToolCall = z.infer<typeof anthropicToolCallSchema>;
+
+/**
+ * The zod schema for multiple Anthropic tool calls
+ */
+export const anthropicToolCallsSchema = z.array(anthropicToolCallSchema);
+
+/**
+ * The JSON schema for multiple Anthropic tool calls
+ */
+export const anthropicToolCallsJSONSchema = zodToJsonSchema(
+  anthropicToolCallsSchema,
+  {
+    removeAdditionalStrategy: "passthrough",
+  }
+);
+
+/**
+ * --------------------------------
+ * Conversion Schemas
+ * --------------------------------
+ */
+
+/**
+ * Parse incoming object as an Anthropic tool call and immediately convert to OpenAI format
+ */
+export const anthropicToOpenAI = anthropicToolCallSchema.transform(
+  (anthropic): OpenAIToolCall => ({
+    id: anthropic.id,
+    function: {
+      name: anthropic.name,
+      arguments: anthropic.input,
+    },
+  })
+);
+
+/**
+ * Parse incoming object as an OpenAI tool call and immediately convert to Anthropic format
+ */
+export const openAIToAnthropic = openAIToolCallSchema.transform(
+  (openai): AnthropicToolCall => ({
+    id: openai.id,
+    type: "tool_use",
+    name: openai.function.name,
+    // REVIEW: anthropic wants a record always, openai wants string, record, or undefined
+    // whats the best way to handle this?
+    input:
+      typeof openai.function.arguments === "string"
+        ? { [openai.function.arguments]: openai.function.arguments }
+        : openai.function.arguments ?? {},
+  })
+);
+
+/**
+ * --------------------------------
+ * Conversion Helpers
+ * --------------------------------
+ */
+
+/**
+ * Union of all tool call formats
+ *
+ * This is useful for functions that need to accept any tool call format
+ */
+export const anyToolCallSchema = z.union([
+  openAIToolCallSchema,
+  anthropicToolCallSchema,
+]);
+
+export type AnyToolCall = z.infer<typeof anyToolCallSchema>;
+
+/**
+ * Detect the provider of a tool call object
+ */
+export const detectProvider = (
+  toolCall: unknown
+): { provider: ModelProvider; validatedToolCall: AnyToolCall } => {
```

Comment:
could you constrain this type so that it's 

```suggestion
): { provider: 'OPENAI'; validatedToolCall: OpenAIToolCall } | Anthrpic stuff => {
```
does that help, seems like it would reduce casting below but would make the return type very unwieldy and hard to maintain

Reply:
I tend to dislike returning unions in this way, and let the consumer do the narrowing on the fields themselves, mostly because it gets hard to read, yeah